activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1312,6 +1312,10 @@ def distinct(value = true)
 
     # Like #distinct, but modifies relation in place.
     def distinct!(value = true) # :nodoc:
+      unless value == true || value == false
```

Comment:
The deprecation should be in the public method, not the private one. 

That said, I'm not convinced this is worth changing. Rails and Ruby are designed to be truthy and falsey. Even if we validate this one, you're still going to have to explain to juniors or anyone new to the language what truthy/falsey means. Also in 99% of cases, anyone calling `distinct` is passing no value at all.

I appreciate you sending a PR, but I'm personally against making this change. Not sure if anyone else on Core feels strongly in favor.

Reply:
The problem with `distinct` is that in SQL, `distinct` takes an argument which is the column you want to unique on (`select distinct name from users`). But in Rails it's different: `User.select(:name).distinct` is fine; `User.distinct(:name)` is wrong.

It's not surprising that (not just junior) developers forget this difference. If you agree that the Rails API is confusing, then I think it's worth doing what we can to make it less confusing.

> Rails and Ruby are designed to be truthy and falsey.

I agree with this in general, but do you think there's valid exceptions (and then do you think this is one of them)?

https://github.com/rails/rails/pull/45229 is another example where technically any truthy argument works, but in practice there's no sensible use case for passing a value other than `true` or `false`.

Reply:
> The deprecation should be in the public method, not the private one. Apps shouldn't call `distinct!` as it's `nodoc`'d.

I was originally raising an `ArgumentError` and [questioned why both shouldn't be consistent with each other](https://github.com/rails/rails/pull/51045#discussion_r1485400123) even if one was private, since it's utilised by the public method.

I left it as is when I updated to warn of a deprecation instead so that it's easier for whoever swaps it for an raise in 7.1.3 to just update the one line, which will raise an error in both as I had it before.

I'm happy to move it to just the public method if we also don't want the private method to raise in 7.1.3.